servers/physics_2d/body_2d_sw.cpp

```diff
@@ -52,9 +57,27 @@ void Body2DSW::update_inertias() {
 			//update tensor for allshapes, not the best way but should be somehow OK. (inspired from bullet)
 			real_t total_area = 0;
 
+			for (int i = 0; i < get_shape_count(); i++) {
+				total_area += get_shape_area(i);
+			}
+
+			center_of_mass_local.x = center_of_mass_local.y = 0;
+
 			for (int i = 0; i < get_shape_count(); i++) {
 
-				total_area += get_shape_aabb(i).get_area();
+				if (is_shape_disabled(i)) {
+					continue;
+				}
+				real_t area = get_shape_area(i);
+				real_t mass = area * this->mass / total_area;
+				Shape2DSW *shape = get_shape(i);
+
+				if (dynamic_cast<ConvexPolygonShape2DSW *>(shape) || dynamic_cast<ConcavePolygonShape2DSW *>(shape)) {
+					Rect2 aabb = shape->get_aabb();
+					center_of_mass_local += mass * get_shape_transform(i).xform(aabb.position + aabb.size / 2);
```

Comment:
It works for rectangular shaped polygons but I don't think this would work flawlessly for irregular shaped polygons.
I think this just needs more treatment. Have a look at these [formulae](https://en.wikipedia.org/wiki/Centroid#Of_a_polygon) for computing polygon's centroid.

Reply:
Whether in 2d or 3d, godot treat polygon as a square when calculating area and inertia. For example :
https://github.com/godotengine/godot/blob/ed2f753fdbacf7d7b4a8f0408dddb19ae2a6707f/servers/physics_2d/shape_2d_sw.cpp#L628-L638
We don't need so much preciousness in fact.

Reply:
I mean in some cases this would be crucial in fact:
![polygon_centroid](https://user-images.githubusercontent.com/17108460/60428293-333c4900-9c01-11e9-94bf-6e2f9b3090a8.png)

So even if Godot doesn't provide such level of precision, a user would like to override computed center of mass, else I'm not sure whether it would be possible to set center of mass by shifting shapes as it would always "mess up" the desired center of mass anyway. The fix should also be translated to 3D case imo, but seems like it might not be that trivial compared to 2D.

Reply:
Physic3D has consideration for center_of_mass already. I'm trying to export a function for custom center_of_mass now, but I'm not sure whether to include it in this PR.